qiskit/primitives/base_sampler.py

```diff
@@ -0,0 +1,155 @@
+    from qiskit.circuit.library import RealAmplitudes
+
+    bell = QuantumCircuit(2, 2)
+    bell.h(0)
+    bell.cx(0, 1)
+    bell.measure(0, 0)
+    bell.measure(1, 1)
+
+    # executes a Bell circuit
+    with Sampler(circuits=bell) as sampler:
+        result = sampler(shots=1000)
+        print([q.binary_probabilities() for q in result.quasi_dists])
+
+    # executes three Bell circuits
+    with Sampler(circuits=[bell, bell, bell], backend=aer_simulator) as sampler:
+        result = sampler(shots=1000)
+        print([q.binary_probabilities() for q in result.quasi_dists])
+
+    # parametrized circuit
+    pqc = QuantumCircuit(2, 2)
+    pqc.compose(RealAmplitudes(num_qubits=2, reps=2), inplace=True)
+    pqc.measure(0, 0)
+    pqc.measure(1, 1)
+
+    theta1 = [0, 1, 1, 2, 3, 5]
+    theta2 = [1, 2, 3, 4, 5, 6]
+
+    with Sampler(circuits=pqc, backend=aer_simulator) as sampler:
+        result1 = sampler([theta1, theta2], shots=3000)
+
+        # result of pqc(theta1)
+        print([q.binary_probabilities() for q in result[0].quasi_dists])
+
+        # result of pqc(theta2)
+        print([q.binary_probabilities() for q in result[1].quasi_dists])
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Optional, Union
+
+from qiskit.circuit import QuantumCircuit, Parameter
+from qiskit.circuit.parametertable import ParameterView
+
+from .sampler_result import SamplerResult
+
+
+class BaseSampler(ABC):
+    """ Sampler base class
+
+    Base class of Sampler that calculates quasi-probabilities of bitstrings from quantum circuits.
+    """
+
+    def __init__(
+        self,
+        circuits: list[QuantumCircuit],
+        parameters: Union[ParameterView, list[Parameter]],
+    ):
+        """
+        Args:
+            circuits (list[QuantumCircuit]): quantum circuits to be executed
+            parameters (Union[ParameterView, list[Parameter]]): parameters of quantum circuits
+        """
+        self._circuits = circuits
+        self._parameters = parameters
+
+    @abstractmethod
+    def __enter__(self):
+        ...
+
+    @abstractmethod
+    def __exit__(self, ex_type, ex_value, trace):
+        ...
+
+    def __init_subclass__(cls):
+        super().__init_subclass__()
+        cls.__call__ = cls.run
+
+    @property
+    def circuits(self) -> list[QuantumCircuit]:
+        """Quantum circuits
+
+        Returns:
+            list[QuantumCircuit]: quantum circuits
+        """
+        return self._circuits
+
+    @property
+    def parameters(self) -> Union[ParameterView, list[Parameter]]:
+        """Parameters of quantum circuits
+
+        Returns:
+            Union[ParameterView, list[Parameter]]: Parameter list of the quantum circuits
+        """
+        return self._parameters
+
+    @abstractmethod
+    def run(
+        self,
+        parameters: Optional[Union[list[float], list[list[float]]]] = None,
+        **run_options,
+    ) -> SamplerResult:
```

Comment:
Shall we introduce an arg (e.g., `circuits: list[int]`) so that users can specify which circuits to evaluate?

Reply:
Yes i think we should